src/include/socket.h

```diff
@@ -312,30 +312,110 @@ static int findInterfaces(char* ifNames, union socketAddress *ifAddrs, int ifNam
+   * NB: The backlog will be silently truncated to the value in /proc/sys/net/core/somaxconn
+   */
+  SYSCHECK(listen(sockfd, 16384), "listen");
+  return ncclSuccess;
+}
+
 static ncclResult_t createListenSocket(int *fd, union socketAddress *localAddr) {
   /* IPv4/IPv6 support */
   int family = localAddr->sa.sa_family;
   int salen = (family == AF_INET) ? sizeof(sockaddr_in) : sizeof(sockaddr_in6);
 
-  /* Create socket and bind it to a port */
-  int sockfd = socket(family, SOCK_STREAM, 0);
-  if (sockfd == -1) {
-    WARN("Net : Socket creation failed : %s", strerror(errno));
-    return ncclSystemError;
+  int sockfd = -1;
+
+  /* Check for a specified port range */
+  char *portRange = getenv("NCCL_PORT_RANGE");
+  unsigned short portMin, portMax;
+  bool portRangeValid = false;
+  if (portRange != NULL) {
+    errno = 0;
+    int ret = sscanf(portRange, "%hu:%hu", &portMin, &portMax);
+    if (ret == 2 && portMax >= portMin) {
+      portRangeValid = true;
+    } else {
+      WARN("Net : invalid NCCL_PORT_RANGE : %s", portRange);
+      return ncclInvalidArgument;
+    }
+    INFO(NCCL_ALL, "Net : Socket port range : %hu:%hu", portMin, portMax);
   }
 
-  if (socketToPort(&localAddr->sa)) {
-    // Port is forced by env. Make sure we get the port.
-    int opt = 1;
-#if defined(SO_REUSEPORT)
-    SYSCHECK(setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &opt, sizeof(opt)), "setsockopt");
-#else
-    SYSCHECK(setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt)), "setsockopt");
-#endif
-  }
+  if (socketToPort(&localAddr->sa) || !portRangeValid) {
+    /* Either a port was set in localAddr, or no port range is specified */
+    sockfd = socket(family, SOCK_STREAM, 0);
+    if (sockfd == -1) {
+      WARN("Net : Socket creation failed : %s", strerror(errno));
+      return ncclSystemError;
+    }
+    bool reuseport = true;
+    NCCLCHECK(bindToPort(sockfd, localAddr, reuseport, NULL));
+  } else {
+    /* Try to find something in the port range */
+    bool portFound = false;
+
+    for (size_t i = portMin; i <= (size_t)portMax; i++) {
+      union socketAddress tryAddr;
+      memcpy(&tryAddr, localAddr, sizeof(tryAddr));
+      if (family == AF_INET) {
+        tryAddr.sin.sin_port = htons((unsigned int)i);
+      } else {
+        tryAddr.sin6.sin6_port = htons((unsigned int)i);
+      }
+
+      sockfd = socket(family, SOCK_STREAM, 0);
+      if (sockfd == -1) {
+        WARN("Net : Socket creation failed : %s", strerror(errno));
+        return ncclSystemError;
+      }
 
-  // localAddr port should be 0 (Any port)
-  SYSCHECK(bind(sockfd, &localAddr->sa, salen), "bind");
+      /* Make sure not to use SO_REUSEPORT because we might bind multiple sockets to the same port */
+      bool reuseport = false;
+      NCCLCHECK(bindToPort(sockfd, &tryAddr, reuseport, &portFound));
+      if (portFound) {
+        break;
+      }
+
+      /* Close the socket and try again */
+      close(sockfd);
```

Comment:
might be cleaner to move the `socket` call inside `bindToPort` because `bindToPort` already alters the socket options

Reply:
Agreed; I chose to separate it this way because of a deficiency in their error handling.  They need a SYSCHECKGOTO() in [checks.h](https://github.com/NVIDIA/nccl/blob/master/src/include/checks.h) to be able to handle this situation properly, if you plan on creating the socket in the same function that you use SYSCHECK.  Of course, they do that exact same thing in the upstream code, but that becomes a problem when we start to tolerate some errors, because we have to properly clean up the socket before trying again.

I opted for the less invasive change (not altering checks.h) over the cleaner change.  An alternative would be to have helper functions for each of `setsockaddr` and `bind`, and pull the call to `listen` out and put it in `createListenSocket` right before `getsockname`?